server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/DKIMSignTest.java

```diff
@@ -0,0 +1,343 @@
+        m7bit.init(mci);
+        m7bit.service(mail);
+
+        ByteArrayOutputStream rawMessage = new ByteArrayOutputStream();
+        mail.getMessage().writeTo(rawMessage);
+
+        MockPublicKeyRecordRetriever mockPublicKeyRecordRetriever = new MockPublicKeyRecordRetriever(
+                "v=DKIM1; k=rsa; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDYDaYKXzwVYwqWbLhmuJ66aTAN8wmDR+rfHE8HfnkSOax0oIoTM5zquZrTLo30870YMfYzxwfB6j/Nz3QdwrUD/t0YMYJiUKyWJnCKfZXHJBJ+yfRHr7oW+UW3cVo9CG2bBfIxsInwYe175g9UjyntJpWueqdEIo1c2bhv9Mp66QIDAQAB;",
+                "selector", "example.com");
+        try {
+            verify(rawMessage, mockPublicKeyRecordRetriever);
+            Assert.fail("Expecting signature to be ignored");
+        } catch (PermFailException e) {
+            // signature ignored, so fail for missing signatures.
+        }
+    }
+
+
+    @Test
+    public void testDKIMSignTime() throws MessagingException, IOException,
+            FailException {
+        String message = "Received: by 10.XX.XX.12 with SMTP id dfgskldjfhgkljsdfhgkljdhfg;\r\n\tTue, 06 Oct 2009 07:37:34 -0700 (PDT)\r\nReturn-Path: <bounce@example.com>\r\nReceived: from example.co.uk (example.co.uk [XX.XXX.125.19])\r\n\tby mx.example.com with ESMTP id dgdfgsdfgsd.97.2009.10.06.07.37.32;\r\n\tTue, 06 Oct 2009 07:37:32 -0700 (PDT)\r\nFrom: apache@bago.org\r\nTo: apache@bago.org\r\n\r\nbody\r\nprova\r\n";
+
+        Mailet mailet = new DKIMSign();
+
+        FakeMailetConfig mci = new FakeMailetConfig("Test", FAKE_MAIL_CONTEXT);
+        mci
+                .setProperty(
+                        "signatureTemplate",
+                        "v=1; t=; s=selector; d=example.com; h=from:to:received:received; a=rsa-sha256; bh=; b=;");
+        mci.setProperty("privateKey", TESTING_PEM);
+
+        mailet.init(mci);
+
+        Mail mail = FakeMail.builder()
+            .mimeMessage(new MimeMessage(Session
+                .getDefaultInstance(new Properties()),
+                new ByteArrayInputStream(message.getBytes())))
+            .build();
+
+        mailet.service(mail);
+
+        Mailet m7bit = new ConvertTo7Bit();
+        m7bit.init(mci);
+        m7bit.service(mail);
+
+        ByteArrayOutputStream rawMessage = new ByteArrayOutputStream();
+        mail.getMessage().writeTo(rawMessage);
+
+        MockPublicKeyRecordRetriever mockPublicKeyRecordRetriever = new MockPublicKeyRecordRetriever(
+                "v=DKIM1; k=rsa; p=MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDYDaYKXzwVYwqWbLhmuJ66aTAN8wmDR+rfHE8HfnkSOax0oIoTM5zquZrTLo30870YMfYzxwfB6j/Nz3QdwrUD/t0YMYJiUKyWJnCKfZXHJBJ+yfRHr7oW+UW3cVo9CG2bBfIxsInwYe175g9UjyntJpWueqdEIo1c2bhv9Mp66QIDAQAB;",
+                "selector", "example.com");
+        verify(rawMessage, mockPublicKeyRecordRetriever);
+
+        List<SignatureRecord> rs = verify(rawMessage, mockPublicKeyRecordRetriever);
+
+        // check we have a valued signatureTimestamp
+        Assert.assertNotNull(rs.get(0).getSignatureTimestamp());
```

Comment:
Moving is one issue, refactoring is an other. Do we want to take time doing this?

(It is an external contribution, we can ask Shrumit to do it...)